org.eclipse.winery.tools.deployablecomponents/src/main/java/org/eclipse/winery/tools/deployablecomponents/crawler/GithubCrawler.java

```diff
@@ -0,0 +1,232 @@
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.net.HttpURLConnection;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.nio.charset.StandardCharsets;
+import java.time.Instant;
+import java.util.ArrayList;
+import java.util.Base64;
+import java.util.List;
+
+public class GithubCrawler implements ICrawlerImplementation {
+
+    private int lastId = 0;
+    private URL baseUrl;
+    private String authEncoded;
+    private boolean authenticated;
+
+    private int rateLimitRemaining = 5000;
+    private Instant rateLimitReset;
+    private int rateLimitRemainingSearch = 30;
+    private Instant rateLimitResetSearch;
+
+    GithubCrawler(String githubName, String oauthToken) {
+        try {
+            baseUrl = new URL("https://api.github.com/");
+            String auth = githubName + ":" + oauthToken;
+            authenticated = !oauthToken.isEmpty();
+            authEncoded = Base64.getEncoder().encodeToString(auth.getBytes(StandardCharsets.UTF_8));
+        } catch (MalformedURLException e) {
+            e.printStackTrace();
+        }
+    }
+
+    public void setStartPoint(int startId) {
+        lastId = startId;
+    }
+
+    public List<Dockerfile> crawlDockerfiles() throws IOException, JSONException {
+        Repository newRepository = getNextRepository();
+        return searchRepositoryForDockerfiles(newRepository);
+    }
+
+    private Repository getNextRepository() throws IOException, JSONException {
+        URL url = new URL(baseUrl + "repositories?since=" + lastId);
+        String responseString = HttpGet(url, true, false);
+
+        JSONArray responseArray = new JSONArray(responseString);
+        JSONObject response = responseArray.getJSONObject(0);
+        lastId = response.getInt("id");
+
+        return parseJsonToRepository(response);
+    }
+
+    private Repository parseJsonToRepository(JSONObject json) throws MalformedURLException, JSONException {
+        String name = json.getString("full_name");
+        String urlString = json.getString("url");
+
+        URL url = new URL(urlString);
+
+        return new Repository(name, url);
+    }
+
+    private String HttpGet(URL url, boolean isInRateLimit, boolean isSearch) throws IOException {
+        if (isInRateLimit && isSearch && rateLimitRemainingSearch <= 0) {
+            sleepSearchRateLimit();
+        }
+        if (isInRateLimit && rateLimitRemaining <= 0) {
+            sleepRateLimit();
+        }
+
+        boolean successful;
+        StringBuilder responseBuilder;
+        HttpURLConnection connection;
+        int fails = 0;
+        // as long as rate limit is not exceeded this do-while is done only once
+        do {
+            if (fails >= 10) {
+                throw new IOException("HTTP request failed " + fails + " times. Request canceled!");
+            }
+            responseBuilder = new StringBuilder();
+            connection = (HttpURLConnection) url.openConnection();
```

Comment:
Don't you also need to close the connection? Please also use the `try(...) {...}` syntax as described below.